### PR TITLE
change(rpc): Select getblocktemplate RPC transactions according to ZIP-317

### LINKED ISSUE #5473: Select `getblocktemplate` transactions according to ZIP-317
## Motivation

We want to select transactions based on ZIP-317.

### Specifications

The block production parts of:
- https://zips.z.cash/zip-0317#block-production
- https://hackmd.io/rgb38M_wSxa9kotyaON1fA (or any actual ZIP updates)

The outdated ZIP-313 might also have some relevant information:
https://zips.z.cash/zip-0313

### Design

When selecting transactions, we should do something similar to `zcashd`, based on ZIPs 317 and 313.

The ZIP editors have recommended including low-fee transactions in blocks with probability `fee / conventional_fee`. Zebra could implement this in the mempool, but that changes relaying as well, which we don't want to do yet. So using the RPC is best for now. The details are in https://hackmd.io/rgb38M_wSxa9kotyaON1fA

We should also check the `sigop_limit` before adding each new transaction to the block.

## Related Tickets

Other parts of transaction selection will be handled by the mempool changes in #5335.

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We want to select transactions for block production according to ZIP-317.

Close #5473.

Depends-On: #5761

### Specifications

https://zips.z.cash/zip-0317#recommended-algorithm-for-block-template-construction

### Designs

Follow the recommended algorithm in the specification.
Use `f32` for efficient calculations and weight storage.

Use a minimum fee weight of `1.0 / 20.0` to simplify the implementation. (This is not in the spec, but changes are allowed because it is not consensus-critical. If the ZIP editors disagree with this, I'll open another PR to fix it.)

Use the same random weight APIs that we use in mempool transaction eviction.

Some parts of this PR are outside the `getblocktemplate-rpcs` feature, because we will probably use them in ticket #5336.

## Solution

- Add a block_production_fee_weight field to VerifiedUnminedTx
- Add a custom Zebra minimum transaction weight for block production
- Randomly select transactions for block production according to ZIP-317

Related Refactors:
- Move getblocktemplate transaction selection into a new zip317 module
- Split the conventional fee check into its own method

This PR has some testing from the existing test vectors and snapshot tests.
The rest of its tests will be implemented as part of ticket #5652.

## Review

We should probably get two reviews on this.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
  - [ ] How do you know it works? Does it have tests?

### CODE REVIEW

File: zebra-rpc/src/methods/get_block_template_rpcs/zip317.rs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    let (conventional_fee_txs, low_fee_txs): (Vec<_>, Vec<_>) = mempool_transactions
+        .into_iter()
+        .partition(VerifiedUnminedTx::pays_conventional_fee);
+
+    // Set up limit tracking
+    let mut selected_txs = Vec::new();
+    let mut remaining_block_sigops = MAX_BLOCK_SIGOPS;
+    let mut remaining_block_bytes: usize = MAX_BLOCK_BYTES.try_into().expect("fits in memory");
+
+    // TODO: split these into separate functions?
```
Comment (teor2345) [inline zebra-rpc/src/methods/get_block_template_rpcs/zip317.rs]: TODO: I might want to split this into separate functions

File: zebra-chain/src/transaction/unmined/zip317.rs [hunk 2]
```diff
@@ -27,6 +24,26 @@ const P2PKH_STANDARD_INPUT_SIZE: usize = 150;
...
+///
+/// If blocks are full, this makes it likely that very low fee transactions
+/// will be mined:
+/// - after approximately 20 blocks delay,
+/// - but before they expire.
+///
+/// Note: Small transactions that pay the legacy ZIP-313 conventional fee have twice this weight.
+/// If blocks are full, they will be mined after approximately 10 blocks delay.
+///
+/// [ZIP-203]: https://zips.z.cash/zip-0203#changes-for-blossom>
```
Comment (oxarbitrage) [inline zebra-chain/src/transaction/unmined/zip317.rs:44]:
```suggestion
/// Note: Small transactions that pay the legacy [ZIP-313] conventional fee have twice this weight.
/// If blocks are full, they will be mined after approximately 10 blocks delay.
///
/// [ZIP-203]: https://zips.z.cash/zip-0203#changes-for-blossom
/// [ZIP-313]: https://zips.z.cash/zip-0313#specification
```

**Comment by teor2345** [comment]
The code in this PR will have further changes after a ZIP update. I suggest we review and merge it based on the current ZIP-317, and then do the update in a separate PR based on ticket #5752.

That will make reviewing the update easier, because the diff will be smaller.

**Comment by teor2345** [comment]
(Oops, pressed the wrong button.)

**Review by oxarbitrage — APPROVED** [review]
Looks good to me.

File: zebra-rpc/src/methods/get_block_template_rpcs/zip317.rs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+                choose_transaction_weighted_random(&low_fee_txs, tx_weights);
+            low_fee_tx_weights = tx_weights;
+
+            // > and add it to the block. If that transaction would exceed the `size_target`
+            // > or the block sigop limit, stop without adding it.
+            if candidate_tx.legacy_sigop_count > remaining_block_sigops
+                || candidate_tx.transaction.size > remaining_block_bytes
+            {
+                // We've exceeded the scaled quota size limit, or the absolute sigop limit
+                break;
```
Comment (arya2) [inline zebra-rpc/src/methods/get_block_template_rpcs/zip317.rs:104]: ~Optional: move the while loop for conventional-fee transactions to its own function and re-use it here.~

Update: nevermind, we should wait for the ZIP update.

**Review by arya2 — APPROVED** [review]
This looks good!

File: zebra-chain/src/transaction/unmined/zip317.rs (hunk 2 above)
Comment (teor2345) [reply zebra-chain/src/transaction/unmined/zip317.rs:44]: Thanks, I've added this to ticket #5752, we might end up rewriting this comment entirely.

**Comment by teor2345** [comment]
@mergifyio refresh

File: zebra-rpc/src/methods/get_block_template_rpcs/zip317.rs (hunk 3 above)
Comment (teor2345) [reply zebra-rpc/src/methods/get_block_template_rpcs/zip317.rs:104]: I thought about this, but the logic was different enough that the abstraction could be confusing.

Let's see what it's like after the update.

**Comment by daira** [comment]
Note that the algorithm has changed, as discussed in ZIP sync. PR here: https://github.com/zcash/zips/pull/650

**Comment by teor2345** [comment]
Please don't worry about restarting the tests in this PR, or merging in main, I'm going to update the code tomorrow anyway.

File: zebra-chain/src/transaction/unmined/zip317.rs (hunk 2 above)
Comment (teor2345) [reply zebra-chain/src/transaction/unmined/zip317.rs:44]: This ended up getting deleted anyway.

**Pull request merged**